cloudsync/smartsync.py

```diff
@@ -0,0 +1,384 @@
+        with self._mutex:
+            for parent_conflict in self.smgr.get_parent_conflicts(ent, REMOTE):  # ALWAYS remote
+                self._sync_one_entry(parent_conflict)
+            return self._sync_one_entry(ent)
+
+    def smart_sync_oid(self, remote_oid):
+        ent: SyncEntry = self.state.smart_sync_oid(remote_oid)
+        if not ent:
+            raise ex.CloudFileNotFoundError(remote_oid)
+        self._smart_sync_ent(ent)
+        return ent[LOCAL].path
+
+    def smart_sync_path(self, path, side):
+        remote_path = self._ensure_path_remote(path, side)
+        try:
+            ents = self.state.smart_sync_path(remote_path)
+        except ex.CloudException as e:
+            self.nmgr.notify_from_exception(SourceEnum.SYNC, e, remote_path)
+            raise
+        for ent in ents:
+            self._smart_sync_ent(ent)
+
+    def smart_listdir_path(self, local_path):
+        """
+        Returns the listdir calculated by the local listdir mixed with the remote sync state, when a file does
+        not exist locally
+        """
+        # assumes that local is efficient and remote is inefficient
+        # read from cache as much as possible remotely, read from disk as much as possible locally
+        # calls local listdir, local listdir overrides cached info, and also update the state with this info too
+        #   remote_oid and is_synced
+        local, remote = self.providers
+        remote_path = self.translate(REMOTE, local_path)
+        local_ents = dict()
+        remote_ents = dict()
+        for dirent in local.listdir_path(local_path):
+            local_ents[dirent.name] = dirent
+        for ent in self.state.smart_listdir_path(remote_path):
+            remote_ents[remote.basename(ent[REMOTE].path)] = ent
+        names = set(local_ents.keys()).union(remote_ents.keys())
+        for name in names:
+            rent = remote_ents.get(name)
+            lent = local_ents.get(name)
+            yield_val = self._ent_to_smartinfo(rent, lent, local.join(local_path, name))
+            yield yield_val
+
+    def _ensure_path_remote(self, path, side) -> str:
+        if side == LOCAL:
+            path = self.translate(REMOTE, path)  # TODO translate the path to remote
+        return path
+
+    def smart_info_path(self, local_path) -> Optional[SmartInfo]:
+        remote_path = self.translate(REMOTE, local_path)
+        ents = self.state.lookup_path(REMOTE, remote_path)
+        if ents:
+            return self._ent_to_smartinfo(ents[0], None, local_path)
+        return None
+
+    def smart_info_oid(self, remote_oid) -> Optional[SmartInfo]:
+        ent = self.state.lookup_oid(REMOTE, remote_oid)
+        if ent:
+            local_path = self.translate(LOCAL, ent[REMOTE].path)
+            return self._ent_to_smartinfo(ent, None, local_path)
+        return None
+
+    def smart_delete_oid(self, remote_oid):
+        ent = self.state.lookup_oid(REMOTE, remote_oid)
```

Comment:
I'm concerned about performance for directory deletes. I think deleting each oid synchronously will kill performance for large directories. I think we need an asynchronous solution